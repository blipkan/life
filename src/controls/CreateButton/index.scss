.button-create {
  display: flex;
  box-sizing: border-box;
  width: 100%;
  margin: 4px 16px;
  transition: all 0.2s ease-in-out;
  &:hover {
    background-color: #432654;
    font-size: 15px;
  }
}
