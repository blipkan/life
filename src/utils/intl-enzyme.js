import React from 'react'
import { IntlProvider } from 'react-intl'
import flatten from 'flat'

const defaultLocale = 'ru'
const locale = defaultLocale

const messages = require(`../../src/messages/${locale}.json`)

export const withIntl = component => {
  return (
    <IntlProvider locale={locale} messages={flatten(messages)}>
      {React.cloneElement(component)}
    </IntlProvider>
  )
}
