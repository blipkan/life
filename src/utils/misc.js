import keys from 'lodash/keys'
import isObject from 'lodash/isObject'

export const makeId = (length = 16) =>
  parseInt(
    Math.ceil(Math.random() * Date.now())
      .toPrecision(length)
      .toString()
      .replace('.', '')
  )

export const uniqArray = (array = []) => [...new Set(array)]

export const compactArray = (array = []) => array.filter(e => e)

export const consoleMessage = (message, messageType) => {
  const consoleMethod = ['error', 'warn', 'log', 'info'].includes(messageType)
    ? messageType
    : 'error'
  console[consoleMethod](message)
}

export const checkKey = (obj, key, errorMessageType, errorMessage) => {
  if (!isObject(obj)) {
    const message =
      errorMessage || `${JSON.stringify(obj)} cannot have key '${key}'`
    consoleMessage(message, errorMessageType)
    return
  }

  let result
  try {
    result = obj[key]
    if (!result) {
      const message =
        errorMessage ||
        `invalid key '${key}' for object: ${JSON.stringify(
          obj
        )}. Available keys: ${keys(obj)}`
      consoleMessage(message, errorMessageType)
    }
  } catch (e) {
    consoleMessage(e.message, errorMessageType)
  }

  return result
}

export const getIntlText = (formatMessage, messages = {}, key) => {
  const message = messages[key]
  return message ? formatMessage(message) : key
}
