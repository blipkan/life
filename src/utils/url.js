/**
 * @description get array from url search parameter
 * @param {string} paramName  parameter name
 * @returns {Array<string>}
 */
export const getValueFromUrlSearch = paramName => {
  const url = new URL(location.href)
  return url.searchParams.get(paramName)
}
