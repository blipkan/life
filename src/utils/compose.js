import { compactArray, uniqArray } from './misc'

export const Delimiter = {
  CLASSNAME: '__',
  MODIFIER: '_',
  ID: '_'
}

const createArrayCompositor = (
  base,
  delimiter = Delimiter.CLASSNAME,
  whiteSpaceAsDelimiter
) => {
  const strings = Array.isArray(base) ? uniqArray(compactArray(base)) : [base]

  const roots = !whiteSpaceAsDelimiter
    ? strings
    : strings.reduce((acc, string) => {
        return acc.concat(...string.split(' '))
      }, [])

  return (...siffixes) => {
    const path = siffixes.join(delimiter)
    return roots.map(
      id => `${id || ''}${id && path ? delimiter : ''}${path || ''}`
    )
  }
}

const createCompositor = (base, delimiter = Delimiter.CLASSNAME) => {
  const arrayCompositor = createArrayCompositor(base, delimiter)
  return (...siffixes) => arrayCompositor(...siffixes).join(' ')
}

const createCompositors = (
  className,
  id,
  {
    classNameDelimiter = Delimiter.CLASSNAME,
    idDelimiter = Delimiter.ID,
    modifierDelimiter = Delimiter.MODIFIER
  } = {}
) => {
  const classNameArrayCompositor = createArrayCompositor(
    className,
    classNameDelimiter,
    true
  )
  const classNameCompositor = (...siffixes) =>
    classNameArrayCompositor(...siffixes).join(' ')

  const modifierCompositor = (modifier, ...siffixes) => {
    const paths = classNameArrayCompositor(...siffixes)
    return paths
      .map(
        path =>
          `${path || ''}${
            modifier && path ? modifierDelimiter : ''
          }${modifier || ''}`
      )
      .join(' ')
  }

  const classNameAndModifierCompositor = (modifier, ...siffixes) => {
    const classNameString = classNameCompositor(...siffixes)
    const modifierString = modifierCompositor(modifier, ...siffixes)
    return `${classNameString} ${modifierString}`
  }

  const idCompositor = createCompositor(id, idDelimiter)

  return {
    id: idCompositor,
    className: classNameCompositor,
    modifier: modifierCompositor,
    classNameAndModifier: classNameAndModifierCompositor
  }
}

export const createCompose = (props = {}, defaults = {}) => {
  const { className, domId } = props
  return createCompositors(
    [defaults.className, className],
    domId || defaults.domId
  )
}
