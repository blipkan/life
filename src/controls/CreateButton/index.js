import React from 'react'
import * as PropTypes from 'prop-types'
import { defineMessages, useIntl } from 'react-intl'

import { createCompose } from 'utils/compose'
import Button from 'controls/common/Button'

import './index.scss'

export const Defaults = {
  className: 'button-create',
  domId: 'button-create'
}

const messages = defineMessages({
  caption: {
    id: 'controls.CreateButton.caption',
    defaultMessage: 'СОТВОРИТЬ'
  },
  hint: {
    id: 'controls.CreateButton.hint',
    defaultMessage: 'Создать новую клетку...'
  }
})

const CreateButton = props => {
  const { onClick } = props

  const { formatMessage } = useIntl()

  const compose = createCompose(props, Defaults)

  const doClick = e => {
    if (typeof onClick === 'function') onClick()
  }

  return (
    <Button
      className={compose.className()}
      domId={compose.id()}
      onClick={doClick}
      caption={formatMessage(messages.caption)}
      hint={formatMessage(messages.hint)}
    />
  )
}

CreateButton.propTypes = {
  onClick: PropTypes.func
}

export default CreateButton
