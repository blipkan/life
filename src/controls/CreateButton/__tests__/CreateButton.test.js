/* eslint-env jest */

import React from 'react'
import { shallow, mount } from 'enzyme'
import { withIntl } from '../../../utils/intl-enzyme'
import { createCompose } from '../../../utils/compose'

import CreateButton, { Defaults } from '../index'

describe('CreateButton', () => {
  const DOMID = 'DOMID'
  const CLASSNAME = 'CLASSNAME'
  const onClick = jest.fn(e => e)

  const Properties = {
    className: CLASSNAME,
    domId: DOMID,
    onClick
  }

  it('render component', () => {
    const component = shallow(withIntl(<CreateButton />))
    expect(component).toMatchSnapshot()
  })

  it('render props', () => {
    const props = { ...Properties }
    const compose = createCompose(props, Defaults)
    const component = mount(withIntl(<CreateButton {...props} />))
    const div = component.find('div')
    expect(div.prop('id')).toBe(compose.id())
    expect(div.prop('className')).toContain(compose.className())
  })

  it('call onClick', () => {
    const props = { ...Properties }

    const component = mount(withIntl(<CreateButton {...props} />))
    const div = component.find('div')
    div.simulate('click')

    expect(onClick).toHaveBeenCalledTimes(1)
  })
})
