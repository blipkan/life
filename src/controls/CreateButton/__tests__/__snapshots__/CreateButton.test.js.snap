// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateButton render component 1`] = `
<ContextProvider
  value={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "defaultRichTextElements": undefined,
      "formatDate": [Function],
      "formatDateTimeRange": [Function],
      "formatDateToParts": [Function],
      "formatDisplayName": [Function],
      "formatList": [Function],
      "formatListToParts": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatNumberToParts": [Function],
      "formatPlural": [Function],
      "formatRelativeTime": [Function],
      "formatTime": [Function],
      "formatTimeToParts": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getDisplayNames": [Function],
        "getListFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralRules": [Function],
        "getRelativeTimeFormat": [Function],
      },
      "locale": "ru",
      "messages": Object {
        "WorldItem.DEAD.description": "или прикидывается",
        "WorldItem.DEAD.title": "Мёртвая",
        "WorldItem.LIFE.description": "Ку-ку!",
        "WorldItem.LIFE.title": "Жизнь",
        "WorldItem.LIVE.description": "и шевелится!",
        "WorldItem.LIVE.title": "Живая",
        "controls.CreateButton.caption": "СОТВОРИТЬ",
        "controls.CreateButton.hint": "Создать новую клетку...",
        "header.controls.cleareButton.caption": "[X]",
        "header.controls.cleareButton.hint": "Очистить историю жизни...",
        "header.titleBar.caption": "Клеточное наполнение",
        "header.titleBar.hint": "Вернуться в начало...",
      },
      "onError": [Function],
      "textComponent": Symbol(react.fragment),
      "timeZone": undefined,
      "wrapRichTextChunksInFragment": undefined,
    }
  }
>
  <CreateButton />
</ContextProvider>
`;
